### PR TITLE
SNOW-1514764: Support Tracing in SPCS runtime

### PR DESCRIPTION
**Pull request by sfc-gh-yuwang**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes #SNOW-1514764

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am modifying authorization mechanisms
   - [ ] I am adding new credentials
   - [ ] I am modifying OCSP code
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

   Please write a short description of how your code change solves the related issue.

4. (Optional) PR for stored-proc connector:

### CODE REVIEW

File: src/snowflake/connector/network.py [hunk 1]
```diff
@@ -487,6 +487,12 @@ def request(
             HTTP_HEADER_ACCEPT: accept_type,
             HTTP_HEADER_USER_AGENT: PYTHON_CONNECTOR_USER_AGENT,
         }
+        try:
+            from opentelemetry.propagate import inject
+
+            inject(headers)
+        except ModuleNotFoundError as e:
+            logger.debug(f"Opentelemtry otel injection failed because of: {e}")
```
Comment (sfc-gh-bdrutu) [inline src/snowflake/connector/network.py:495]: Not sure if the log is useful or not...

**Review by sfc-gh-bdrutu — APPROVED** [review]

**Comment by sfc-gh-yuwang** [comment]
@sfc-gh-bdrutu Hi, can you give me a brief explanation why this change has to be in python connector? Is there other ways that we can do this?

File: src/snowflake/connector/network.py (hunk 1 above)
Comment (sfc-gh-yuwang) [reply src/snowflake/connector/network.py:495]: I think it is better than put a `pass` here

File: src/snowflake/connector/network.py [hunk 2]
```diff
@@ -487,6 +487,12 @@ def request(
             HTTP_HEADER_ACCEPT: accept_type,
             HTTP_HEADER_USER_AGENT: PYTHON_CONNECTOR_USER_AGENT,
         }
+        try:
+            from opentelemetry.propagate import inject
+
+            inject(headers)
```
Comment (sfc-gh-aling) [inline src/snowflake/connector/network.py:493]: 1. `request` method is on the critical path of query execution, what's the overhead of opentelemetry inject added here, e.g., increase in the payload, potential perf regression? also would try-except cause perf regression? I think we need to test perf of this change, and see the payload change
2. should we allow customers to opt out open telemetry collection?
3. how would backend react to request with new header? do we need any backend change for compatibility?

Comment (sfc-gh-yuwang) [reply src/snowflake/connector/network.py:493]: @sfc-gh-bdrutu I think some of this question are answered in slack chat, but can you still take a look?

Comment (sfc-gh-aling) [reply src/snowflake/connector/network.py:493]: just synced up with @sfc-gh-yuwang , I think once we get perf tested (can be as simple as performing "select 1" for 1 million times) and there is no regression, then let's merge it to unblock SPCS scenario first.

Comment (sfc-gh-bdrutu) [reply src/snowflake/connector/network.py:493]: We will add one header "traceparent" which has 55 characters value (always). So the overhead shouldn't be that significant especially it does not happen if the user doesn't configure opentelemetry. Not sure about performance of try/catch.

> how would backend react to request with new header, do we need any backend change?

We are taking care of that.

**Review by sfc-gh-aling — APPROVED** [review]

**Pull request merged**